blockchain/utxoentry.go

```diff
@@ -0,0 +1,209 @@
+	// utxoFlags.  This approach is used in order to reduce memory usage since
+	// there will be a lot of these in memory.
+	packedFlags utxoFlags
+}
+
+// isModified returns whether or not the output has been modified since it was
+// loaded.
+func (entry *UtxoEntry) isModified() bool {
+	return entry.packedFlags&utxoFlagModified == utxoFlagModified
+}
+
+// IsCoinBase returns whether or not the output was contained in a coinbase
+// transaction.
+func (entry *UtxoEntry) IsCoinBase() bool {
+	return entry.packedFlags&utxoFlagCoinBase == utxoFlagCoinBase
+}
+
+// IsSpent returns whether or not the output has been spent based upon the
+// current state of the unspent transaction output view it was obtained from.
+func (entry *UtxoEntry) IsSpent() bool {
+	return entry.packedFlags&utxoFlagSpent == utxoFlagSpent
+}
+
+// HasExpiry returns whether or not the output was contained in a transaction
+// that included an expiry.
+func (entry *UtxoEntry) HasExpiry() bool {
+	return entry.packedFlags&utxoFlagHasExpiry == utxoFlagHasExpiry
+}
+
+// BlockHeight returns the height of the block containing the output.
+func (entry *UtxoEntry) BlockHeight() int64 {
+	return int64(entry.blockHeight)
+}
+
+// BlockIndex returns the index of the transaction that the output is contained
+// in.
+func (entry *UtxoEntry) BlockIndex() uint32 {
+	return entry.blockIndex
+}
+
+// TransactionType returns the type of the transaction that the output is
+// contained in.
+func (entry *UtxoEntry) TransactionType() stake.TxType {
+	txType := (entry.packedFlags & utxoFlagTxTypeBitmask) >> utxoFlagTxTypeShift
+	return stake.TxType(txType)
+}
+
+// Spend marks the output as spent.  Spending an output that is already spent
+// has no effect.
+func (entry *UtxoEntry) Spend() {
+	// Nothing to do if the output is already spent.
+	if entry.IsSpent() {
+		return
+	}
+
+	// Mark the output as spent and modified.
+	entry.packedFlags |= utxoFlagSpent | utxoFlagModified
+}
+
+// Amount returns the amount of the output.
+func (entry *UtxoEntry) Amount() int64 {
+	return entry.amount
+}
+
+// PkScript returns the public key script for the output.
+func (entry *UtxoEntry) PkScript() []byte {
+	return entry.pkScript
+}
+
+// ScriptVersion returns the public key script version for the output.
+func (entry *UtxoEntry) ScriptVersion() uint16 {
+	return entry.scriptVersion
+}
+
+// TicketMinimalOutputs returns the minimal outputs for the ticket transaction
+// that the output is contained in.  Note that the ticket minimal outputs are
+// only stored in ticket submission outputs and nil will be returned for all
+// other output types.
+func (entry *UtxoEntry) TicketMinimalOutputs() []*stake.MinimalOutput {
+	if entry.ticketMinOuts == nil {
+		return nil
+	}
+
+	minOuts, _ := deserializeToMinimalOutputs(entry.ticketMinOuts.data)
+	return minOuts
+}
+
+// Clone returns a copy of the utxo entry.  It performs a deep copy for any
+// fields that are mutable.
```

Comment:
Was reviewing and while not directly related to this PR, I noticed this comment isn't true. Specifically, the pkScript and ticketMinOuts are **not** actually deep-copied, rather only their reference is copied.

Reply:
Pretty sure I pointed out that same thing on another PR.

Reply:
Ya, I had added the stipulation `for any fields that are mutable.`  I could update it to clarify that specifically:
```
// Clone returns a copy of the utxo entry.  It performs a deep copy for any
// fields that are mutable.  Specifically, the script and ticket minimal outputs
// are NOT deep copied since they are immutable.
```

Reply:
That sounds pretty clear to me.  I like it.

Reply:
Added an extra commit to improve that comment as above.